lib/gui/modules/child-writer.js

```diff
@@ -0,0 +1,130 @@
+const errors = require('../../shared/errors')
+const writer = require('../../cli/writer')
+
+ipc.config.id = process.env.IPC_CLIENT_ID
+ipc.config.socketRoot = process.env.IPC_SOCKET_ROOT
+ipc.config.silent = false
+
+// > If set to 0, the client will NOT try to reconnect.
+// See https://github.com/RIAEvangelist/node-ipc/
+//
+// The purpose behind this change is for this process
+// to emit a "disconnect" event as soon as the GUI
+// process is closed, so we can kill this process as well.
+ipc.config.stopRetrying = 0
+
+const IPC_SERVER_ID = process.env.IPC_SERVER_ID
+
+/**
+ * @summary Send a log debug message to the IPC server
+ * @function
+ * @private
+ *
+ * @param {String} message - message
+ *
+ * @example
+ * log('Hello world!')
+ */
+const log = (message) => {
+  ipc.of[IPC_SERVER_ID].emit('log', message)
+}
+
+/**
+ * @summary Terminate the child writer process
+ * @function
+ * @private
+ *
+ * @param {Number} [code=0] - exit code
+ *
+ * @example
+ * terminate(1)
+ */
+const terminate = (code) => {
+  ipc.disconnect(IPC_SERVER_ID)
+  process.exit(code || EXIT_CODES.SUCCESS)
+}
+
+/**
+ * @summary Handle a child writer error
+ * @function
+ * @private
+ *
+ * @param {Error} error - error
+ *
+ * @example
+ * handleError(new Error('Something bad happened!'))
+ */
+const handleError = (error) => {
+  ipc.of[IPC_SERVER_ID].emit('error', errors.toJSON(error))
+  terminate(EXIT_CODES.GENERAL_ERROR)
+}
+
+ipc.connectTo(IPC_SERVER_ID, () => {
+  process.once('uncaughtException', handleError)
+
+  // Gracefully exit on the following cases. If the parent
+  // process detects that child exit successfully but
+  // no flashing information is available, then it will
+  // assume that the child died halfway through.
+
+  process.once('SIGINT', () => {
+    terminate(EXIT_CODES.SUCCESS)
+  })
+  process.once('SIGTERM', () => {
+    terminate(EXIT_CODES.SUCCESS)
+  })
+
+  // The IPC server failed. Abort.
+  ipc.of[IPC_SERVER_ID].on('error', () => {
+    terminate(EXIT_CODES.SUCCESS)
+  })
+
+  // The IPC server was disconnected. Abort.
+  ipc.of[IPC_SERVER_ID].on('disconnect', () => {
+    terminate(EXIT_CODES.SUCCESS)
+  })
+
+  ipc.of[process.env.IPC_SERVER_ID].on('connect', () => {
+    log(`Successfully connected to IPC server: ${IPC_SERVER_ID}, socket root ${ipc.config.socketRoot}`)
+    log(`Image: ${process.env.OPTION_IMAGE}`)
+    log(`Device: ${process.env.OPTION_DEVICE}`)
+
+    // These come as strings containing 0 or 1. We need to convert
+    // them to numbers first, and then to booleans, otherwise something
+    // like `Boolean('0')` will be `true`
+    const unmountOnSuccess = Boolean(Number(process.env.OPTION_UNMOUNT))
+    const validateWriteOnSuccess = Boolean(Number(process.env.OPTION_VALIDATE))
+    log(`Umount on success: ${unmountOnSuccess}`)
+    log(`Validate on success: ${validateWriteOnSuccess}`)
+
+    writer.writeImage(process.env.OPTION_IMAGE, process.env.OPTION_DEVICE, {
```

Comment:
Should we pass this over IPC instead of through env vars? If we want to add persistence and stopping it might make more sense.

Reply:
Yeah, that's next step (otherwise this PR will keep growing). I want to get most of this merged, and will send and follow up PR that elevates on long-living process at startup that does scanning and flashing using IPC commands.